src/lib/token_list_gen.ts

```diff
@@ -67,87 +77,161 @@ export const generateTokenList = async (
       symbol,
       decimals,
       extensions: {
-        l1Address: token.id,
-        l2GatewayAddress,
-        l1GatewayAddress: l2ToL1GatewayAddresses[l2GatewayAddress],
-      },
+        bridgeInfo: {
+          [l1Network.chainID]: {
+            tokenAddress: token.l1TokenAddr,
+            originBridgeAddress: l2GatewayAddress,
+            destBridgeAddress: l2ToL1GatewayAddresses[l2GatewayAddress.toLowerCase()]
+          }
+        }
+  
+      }
     };
     if (logoUris[i]) {
       arbTokenInfo = { ...{ logoURI: logoUris[i] }, ...arbTokenInfo };
     } else {
-      console.log('no logo uri for ',token.id, symbol);
+      console.log('no logo uri for ',token.l1TokenAddr, symbol);
       
     }
 
     return arbTokenInfo;
   }).filter((tokenInfo: ArbTokenInfo)=>{
-    return tokenInfo.extensions.l2GatewayAddress !== "0x0000000000000000000000000000000000000001" 
+    return tokenInfo.extensions.bridgeInfo[l1Network.chainID].originBridgeAddress !== "0x0000000000000000000000000000000000000001" 
   })
-  //   @ts-ignore
   tokenList.sort((a, b) => (a.symbol < b.symbol ? -1 : 1));
-  const version = prevVersion || {
-    major: 1,
-    minor: 0,
-    patch: 0,
-  }
+
+  const version = (()=>{
+    if(prevArbTokenList){
+      // @ts-ignore
+      let versionBump = minVersionBump(prevArbTokenList.tokens, tokenList)
+
+      // tmp: library doesn't nicely handle patches (for extensions object)
+      if(versionBump === VersionUpgrade.PATCH){
+        versionBump = VersionUpgrade.NONE
+      }
+      return nextVersion(prevArbTokenList.version, versionBump)  
+    }
+    return  {
+      major: 1,
+      minor: 0,
+      patch: 0,
+    }
+  })()
+
   const arbTokenList: ArbTokenList = {
-    name: sanitizeString(`Arbed ${name}`.slice(0,19)),
+    name: listNameToArbifiedListName(name),
     timestamp: new Date().toISOString(),
     version,
     tokens: tokenList,
-    logoURI: mainLogoUri // todo: handle undefined
+    logoURI: mainLogoUri
   };
   const res = validateTokenList(arbTokenList);
   if(!res){
-    console.log(arbTokenList);    
-    throw new Error("New token list invalid!")
+    console.log("Token list invalid — let's try to see why:");
+    while(arbTokenList.tokens.length > 0){
+      const candidateToken = arbTokenList.tokens.pop()
+      const res = validateTokenList(arbTokenList);
```

Comment:
would it be faster to instead validate a token list that only includes the candidate token that you popped?

Reply:
pretty much instant either way